samza-api/src/main/java/org/apache/samza/startpoint/Startpoint.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.startpoint;
+
+import com.google.common.base.Objects;
+import java.time.Instant;
+import org.apache.samza.annotation.InterfaceStability;
+import org.apache.samza.system.SystemStreamPartition;
+
+
+/**
+ * Startpoint represents a position in a stream partition.
+ */
+@InterfaceStability.Unstable
+public abstract class Startpoint {
+
+  private final Long createdTimestamp;
+
+  protected Startpoint() {
```

Comment:
I was thinking of the case where someone writes the Startpoint, and then resets the job, and then the Startpoint gets read. Then the write time is going to be significantly before the read time.
In my opinion, it would be nice to avoid setters, since it is slightly harder to ensure that it gets called correctly. If you require it as a constructor param, then the caller of the constructor could pass in Instant.now(). Maybe one way to also be consistent with @vjagadish1989 's suggestion is to not have a Startpoint.copyWithStoredAt and directly use the Startpoint (including the createdTimestamp in it) passed to StartpointManager.writeStartpointForTask.